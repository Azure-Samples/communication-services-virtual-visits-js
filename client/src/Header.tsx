// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { Stack, Text, ThemeContext, createTheme } from '@fluentui/react';
import { Theme } from '@fluentui/theme';
import { WaffleMenu } from './WaffleMenu';
import { headerContainerStyles, headerTextStyles } from './styles/Header.styles';

/**
 * @interface HeaderProps The properties that Header uses.
 * @member parentid The ID of the element that will be used to contain the panel. This is required so that the panel does not cover the entire height of the screen.
 */
interface HeaderProps {
  companyName: string;
  parentid: string;
}

export const Header = (props: HeaderProps): JSX.Element => {
  const { companyName, parentid } = props;

  return (
    <ThemeContext.Consumer>
      {(theme: Theme | undefined) => {
        if (theme === undefined) {
          theme = createTheme();
        }
        return (
          <Stack styles={headerContainerStyles(theme)} verticalAlign="center" horizontal>
            <WaffleMenu parentid={parentid} />
            <Text styles={headerTextStyles(theme)}>{companyName}</Text>
          </Stack>
        );
      }}
    </ThemeContext.Consumer>
  );
};
