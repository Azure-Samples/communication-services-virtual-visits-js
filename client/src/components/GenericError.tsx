// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { Icon, Stack, Text } from '@fluentui/react';
import { errorIconStackStyle, errorTitleStyle } from '../styles/GenericError.styles';

interface GenericErrorProps {
  statusCode: any | undefined;
}

export const GenericError = (props: GenericErrorProps): JSX.Element => {
  const { statusCode } = props;
  const error = genericErrors[statusCode] || genericErrors['default'];

  return (
    <Stack id="generic-error">
      <Text as={'h2'} className={errorIconStackStyle}>
        <Icon iconName="error" />
      </Text>
      <Text as={'h3'} className={errorTitleStyle}>
        {error.title}
      </Text>
      <Text>{error.description}</Text>
    </Stack>
  );
};

const genericErrors = {
  default: {
    title: 'Unknown error',
    description: 'An unknown error has occurred.'
  },
  404: {
    title: '404: Page not found',
    description: 'The requested page could not be found.'
  },
  500: {
    title: '500: Internal server error',
    description: 'The server has encountered an error. Refresh the page to try again.'
  }
};
