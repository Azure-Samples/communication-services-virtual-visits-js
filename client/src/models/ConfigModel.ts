// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { Theme } from '@fluentui/theme';

export type PostCallSurveyType = 'msforms' | 'custom' | 'onequestionpoll';
export interface MSFormsSurveyOptions {
  surveyUrl: string;
}
export interface CustomSurveyOptions {
  surveyUrl: string;
}

export type OneQuestionPollType = 'likeOrDislike' | 'rating' | 'text';

export interface OneQuestionPollOptions {
  title?: string;
  prompt: string;
  pollType: OneQuestionPollType;
  answerPlaceholder?: string;
  saveButtonText: string;
}

export interface PostCallConfig {
  survey: {
    type: PostCallSurveyType;
    options: MSFormsSurveyOptions | CustomSurveyOptions | OneQuestionPollOptions;
  };
}
export interface AppConfigModel {
  communicationEndpoint: string;
  microsoftBookingsUrl: string;
  chatEnabled: boolean;
  screenShareEnabled: boolean;
  companyName: string;
  theme: Theme;
  waitingTitle: string;
  waitingSubtitle: string;
  logoUrl: string;
  postCall?: PostCallConfig;
}
