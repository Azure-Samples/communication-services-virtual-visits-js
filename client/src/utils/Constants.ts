// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

export const BOOKINGS_README_URL = 'https://aka.ms/virtual-appointments-sample-bookings';
export const BOOKINGS_SPECIMEN_URL =
  'https://microsoftbookings.azurewebsites.net/?organization=financialservices&UICulture=en-US';
