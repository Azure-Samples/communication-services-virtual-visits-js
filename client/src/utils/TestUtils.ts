// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import {
  AdapterError,
  CallAdapter,
  CallAdapterState,
  CallWithChatAdapter,
  CallWithChatAdapterState
} from '@azure/communication-react';
import { act } from '@testing-library/react';

export const runFakeTimers = async (): Promise<void> => {
  await act(async () => {
    jest.useFakeTimers();
    jest.runAllTimers();
  });
};

// Inspired by ui-library unit test: https://aka.ms/AAfyg3m
export const createMockCallWithChatAdapter = (): CallWithChatAdapter => {
  const callWithChatAdapter = {} as CallWithChatAdapter;
  callWithChatAdapter.onStateChange = jest.fn();
  callWithChatAdapter.offStateChange = jest.fn();
  callWithChatAdapter.askDevicePermission = jest.fn();
  callWithChatAdapter.queryCameras = jest.fn();
  callWithChatAdapter.queryMicrophones = jest.fn();
  callWithChatAdapter.querySpeakers = jest.fn();
  callWithChatAdapter.on = jest.fn(); // allow for direct subscription to the state of the call-with-chat adapter
  callWithChatAdapter.off = jest.fn(); // Allow for direct un-subscription to the state of the call-with-chat adapter
  callWithChatAdapter.getState = jest.fn(
    (): CallWithChatAdapterState => ({
      page: 'lobby',
      isLocalPreviewMicrophoneEnabled: false,
      userId: { kind: 'communicationUser', communicationUserId: 'test' },
      displayName: 'test',
      devices: {
        isSpeakerSelectionAvailable: false,
        cameras: [],
        microphones: [],
        speakers: [],
        unparentedViews: []
      },
      isTeamsCall: true,
      call: undefined,
      chat: undefined,
      latestCallErrors: { test: new Error() as AdapterError },
      latestChatErrors: { test: new Error() as AdapterError }
    })
  );
  return callWithChatAdapter;
};

export const createMockCallAdapter = (): CallAdapter => {
  const callAdapter = {} as CallAdapter;
  callAdapter.onStateChange = jest.fn();
  callAdapter.offStateChange = jest.fn();
  callAdapter.askDevicePermission = jest.fn();
  callAdapter.queryCameras = jest.fn();
  callAdapter.queryMicrophones = jest.fn();
  callAdapter.querySpeakers = jest.fn();
  callAdapter.on = jest.fn(); // allow for direct subscription to the state of the call adapter
  callAdapter.off = jest.fn(); // Allow for direct un-subscription to the state of the call adapter
  callAdapter.getState = jest.fn(
    (): CallAdapterState => ({
      page: 'lobby',
      isLocalPreviewMicrophoneEnabled: false,
      userId: { kind: 'communicationUser', communicationUserId: 'test' },
      displayName: 'test',
      call: undefined,
      devices: {
        isSpeakerSelectionAvailable: false,
        cameras: [],
        microphones: [],
        speakers: [],
        unparentedViews: []
      },
      isTeamsCall: false,
      latestErrors: { test: new Error() as AdapterError }
    })
  );
  return callAdapter;
};

export const createMockStatefulCallClient = () => {
  return { createCallAgent: () => '' };
};

export const createMockStatefulChatClient = () => {
  return { startRealtimeNotifications: () => '', getChatThreadClient: () => '' };
};

export const createMockCallWithChatComposite = () => {
  return 'hello';
};

export const createMockCallComposite = () => {
  return 'hello';
};
