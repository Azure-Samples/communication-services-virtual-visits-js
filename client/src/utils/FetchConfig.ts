// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { AppConfigModel } from '../models/ConfigModel';
import { generateTheme } from './ThemeGenerator';
import createError from 'http-errors';

export async function fetchConfig(): Promise<AppConfigModel | undefined> {
  const response = await fetch('/api/config', { method: 'GET' });
  if (response.status !== 200) {
    throw new createError(
      response.status,
      `Unable to retrieve config. Status: ${response.status}. Message: ${response.statusText}.`
    );
  }

  const responseContent = await response.text();
  const config = JSON.parse(responseContent);
  if (!config) {
    throw new Error(`Received empty config response.`);
  }

  const appConfigModel: AppConfigModel = {
    ...config,
    theme: generateTheme(config.colorPalette)
  };

  return appConfigModel;
}
