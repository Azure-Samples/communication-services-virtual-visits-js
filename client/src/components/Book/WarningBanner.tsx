// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { MessageBar, MessageBarType, Text, Link } from '@fluentui/react';
import { BOOKINGS_README_URL } from '../../utils/Constants';

const WarningBanner = (): JSX.Element => {
  return (
    <MessageBar messageBarType={MessageBarType.warning} isMultiline={false} dismissButtonAriaLabel="Close">
      <WarningMessage />
    </MessageBar>
  );
};

// Extracting to a component since MessageBar is not rendering child components in tests
export const WarningMessage = (): JSX.Element => {
  return (
    <>
      <Text>This Booking page is a specimen and will not create appointments or send reminders.</Text>
      <Link href={BOOKINGS_README_URL} target="_blank" underline>
        <Text>Check out the sample readme.</Text>
      </Link>
    </>
  );
};

export default WarningBanner;
