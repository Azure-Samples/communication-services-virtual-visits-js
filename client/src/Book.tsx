// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { LayerHost, Spinner, Stack, ThemeProvider } from '@fluentui/react';
import { backgroundStyles, fullSizeStyles } from './styles/Common.styles';
import { Header } from './Header';
import './styles/Common.css';
import { fetchConfig } from './utils/FetchConfig';
import { AppConfigModel } from './models/ConfigModel';
import { GenericError } from './components/GenericError';
import { useEffect, useState } from 'react';
import { BookingsPage } from './components/book/BookingsPage';
import { NoSchedulingPage } from './components/book/NoSchedulingPage';
import GenericContainer from './components/GenericContainer';

export const Book = (): JSX.Element => {
  const [config, setConfig] = useState<AppConfigModel | undefined>(undefined);
  const [error, setError] = useState<any | undefined>(undefined);

  const PARENT_ID = 'BookMeetingSection';

  useEffect(() => {
    const fetchData = async (): Promise<void> => {
      try {
        const fetchedConfig = await fetchConfig();
        setConfig(fetchedConfig);
      } catch (error) {
        console.error(error);
        setError(error);
      }
    };

    fetchData();
  }, []);

  if (config) {
    return (
      <ThemeProvider theme={config.theme} style={{ height: '100%' }}>
        <Stack styles={backgroundStyles(config.theme)}>
          <Header companyName={config.companyName} parentid={PARENT_ID} />
          {config.microsoftBookingsUrl ? (
            <LayerHost
              id={PARENT_ID}
              style={{
                position: 'relative',
                height: '100%'
              }}
            >
              <BookingsPage config={config} />
            </LayerHost>
          ) : (
            <GenericContainer layerHostId={PARENT_ID} theme={config.theme}>
              <NoSchedulingPage />
            </GenericContainer>
          )}
        </Stack>
      </ThemeProvider>
    );
  }

  if (error) {
    return <GenericError statusCode={error.statusCode} />;
  }

  return <Spinner styles={fullSizeStyles} />;
};
