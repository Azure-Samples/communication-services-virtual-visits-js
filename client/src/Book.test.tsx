// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { Spinner } from '@fluentui/react';
import { mount } from 'enzyme';
import { Book } from './Book';
import { Header } from './Header';
import { GenericError } from './components/GenericError';
import { AppConfigModel } from './models/ConfigModel';
import * as FetchConfig from './utils/FetchConfig';
import { runFakeTimers } from './utils/TestUtils';
import { generateTheme } from './utils/ThemeGenerator';
import * as renderer from 'react-test-renderer';
import { BOOKINGS_SPECIMEN_URL } from './utils/Constants';
import WarningBanner from './components/book/WarningBanner';

describe('Book', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error').mockImplementation();
  });

  it('should render loading spinner when config is not loaded', async () => {
    const fetchConfigSpy = jest.spyOn(FetchConfig, 'fetchConfig');
    fetchConfigSpy.mockReturnValue(Promise.resolve(undefined));

    const book = await mount(<Book />);

    await runFakeTimers();

    book.update();

    const spinners = book.find(Spinner);
    const headers = book.find(Header);

    expect(spinners.length).toBe(1);
    expect(headers.length).toBe(0);
  });

  it('renders an generic error UI when config throws an error', async () => {
    const fetchConfigSpy = jest.spyOn(FetchConfig, 'fetchConfig');
    fetchConfigSpy.mockImplementation(
      async (): Promise<AppConfigModel | undefined> => {
        throw new Error('test error');
      }
    );

    const book = await mount(<Book />);

    await runFakeTimers();

    book.update();

    const spinners = book.find(Spinner);
    const genericError = book.find(GenericError);

    expect(spinners.length).toBe(0);
    expect(genericError.length).toBe(1);
  });

  it('should render header and bookings iframe when config is loaded and bookings link is not empty', async () => {
    const mockBookingsUrl = 'https://example.org';
    const fetchConfigSpy = jest.spyOn(FetchConfig, 'fetchConfig');
    fetchConfigSpy.mockReturnValue(
      Promise.resolve({
        communicationEndpoint: 'endpoint=test_endpoint;',
        microsoftBookingsUrl: mockBookingsUrl,
        chatEnabled: true,
        screenShareEnabled: true,
        companyName: '',
        theme: generateTheme('#FFFFFF'),
        waitingTitle: '',
        waitingSubtitle: '',
        logoUrl: ''
      } as AppConfigModel)
    );

    const book = await mount(<Book />);

    await runFakeTimers();

    book.update();

    const spinners = book.find(Spinner);
    const headers = book.find(Header);
    const iframes = book.find('iframe');
    const warningBanner = book.find(WarningBanner);

    expect(spinners.length).toBe(0);
    expect(warningBanner.length).toBe(0);
    expect(headers.length).toBe(1);
    expect(iframes.length).toBe(1);
    expect(iframes.first().props().src).toBe(mockBookingsUrl);
  });

  it('should render header and no scheduling info when config is loaded and bookings link is empty', async () => {
    const fetchConfigSpy = jest.spyOn(FetchConfig, 'fetchConfig');
    fetchConfigSpy.mockReturnValue(
      Promise.resolve({
        communicationEndpoint: 'endpoint=test_endpoint;',
        microsoftBookingsUrl: '',
        chatEnabled: true,
        screenShareEnabled: true,
        companyName: '',
        theme: generateTheme('#FFFFFF'),
        waitingTitle: '',
        waitingSubtitle: '',
        logoUrl: ''
      } as AppConfigModel)
    );

    const book = await mount(<Book />);

    await runFakeTimers();

    book.update();

    const spinners = book.find(Spinner);
    const headers = book.find(Header);
    const iframes = book.find('iframe');
    const warningBanner = book.find(WarningBanner);

    expect(spinners.length).toBe(0);
    expect(warningBanner.length).toBe(0);
    expect(headers.length).toBe(1);
    expect(iframes.length).toBe(0);
  });

  it('should render warning banner if using specimen Bookings page', async () => {
    const fetchConfigSpy = jest.spyOn(FetchConfig, 'fetchConfig');
    fetchConfigSpy.mockReturnValue(
      Promise.resolve({
        communicationEndpoint: 'endpoint=test_endpoint;',
        microsoftBookingsUrl: BOOKINGS_SPECIMEN_URL,
        chatEnabled: true,
        screenShareEnabled: true,
        companyName: '',
        theme: generateTheme('#FFFFFF'),
        waitingTitle: '',
        waitingSubtitle: '',
        logoUrl: ''
      } as AppConfigModel)
    );

    const book = await mount(<Book />);

    await runFakeTimers();
    book.update();

    const warningBanner = book.find(WarningBanner);

    const spinners = book.find(Spinner);
    const headers = book.find(Header);
    const iframes = book.find('iframe');

    expect(spinners.length).toBe(0);
    expect(warningBanner.length).toBe(1);
    expect(headers.length).toBe(1);
    expect(iframes.length).toBe(1);
    expect(iframes.first().props().src).toBe(BOOKINGS_SPECIMEN_URL);
  });

  it('should match snapshot when bookings link is empty', async () => {
    const fetchConfigSpy = jest.spyOn(FetchConfig, 'fetchConfig');
    fetchConfigSpy.mockReturnValue(
      Promise.resolve({
        communicationEndpoint: 'endpoint=test_endpoint;',
        microsoftBookingsUrl: '',
        chatEnabled: true,
        screenShareEnabled: true,
        companyName: '',
        theme: generateTheme('#FFFFFF'),
        waitingTitle: '',
        waitingSubtitle: '',
        logoUrl: ''
      } as AppConfigModel)
    );
    const book = renderer.create(<Book />);
    await runFakeTimers();

    expect(book.toJSON()).toMatchSnapshot();
  });
});
