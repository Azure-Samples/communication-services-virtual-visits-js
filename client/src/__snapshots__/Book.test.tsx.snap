// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book should match snapshot when bookings link is empty 1`] = `
<div
  className="body-118"
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="ms-Stack css-119"
  >
    <div
      className="ms-Stack css-120"
    >
      <div>
        <button
          aria-label="Menu Button"
          className="ms-Button ms-Button--icon root-121"
          data-is-focusable={true}
          id="waffle-menu"
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          type="button"
        >
          <span
            className="ms-Button-flexContainer flexContainer-122"
            data-automationid="splitbuttonprimary"
          >
            <i
              aria-hidden={true}
              className="ms-Icon root-105 ms-Button-icon icon-124"
              data-icon-name="Waffle"
              style={
                Object {
                  "fontFamily": undefined,
                }
              }
            />
          </span>
        </button>
      </div>
      <span
        className="css-128"
      >
        
      </span>
    </div>
    <div
      className="ms-LayerHost"
      id="BookMeetingSection"
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "height": "100%",
          "justifyContent": "center",
          "position": "relative",
        }
      }
    >
      <div
        className="ms-Stack css-129"
      >
        <div
          className="ms-Stack css-130"
        >
          <div
            className="ms-Stack css-131"
          >
            <span
              className="css-132"
            >
              No scheduling configured
            </span>
            <span
              className="css-133"
            >
              This sample does not have a Microsoft Bookings page configured.
            </span>
            <span
              className="css-134"
            >
              Frequently asked questions
            </span>
            <div
              className="ms-Stack css-135"
            >
              <a
                className="ms-Link root-136"
                href="https://aka.ms/virtual-appointments-sample-bookings"
                onClick={[Function]}
                tabIndex={0}
                target="_blank"
              >
                <div
                  className="ms-Stack css-137"
                >
                  <div
                    className="ms-StackItem css-138"
                    style={
                      Object {
                        "fontSize": ".9375rem",
                        "fontWeight": "400",
                        "letterSpacing": "-0.015rem",
                        "lineHeight": "1rem",
                        "paddingRight": ".5rem",
                        "textDecoration": "underline",
                        "verticalAlign": "bottom",
                      }
                    }
                  >
                    How do I change my Microsoft Bookings page URL?
                  </div>
                  <div
                    className="ms-StackItem css-138"
                    style={
                      Object {
                        "textDecoration": "none",
                      }
                    }
                  >
                    <span
                      style={
                        Object {
                          "alignItems": "center",
                          "display": "flex",
                        }
                      }
                    >
                      <i
                        aria-hidden={true}
                        className="root-105"
                        data-icon-name="OpenInNewWindow"
                      />
                    </span>
                  </div>
                </div>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
